php/class-customize-snapshot-manager.php

```diff
@@ -0,0 +1,559 @@
+	 * Save a snapshot.
+	 *
+	 * @param \WP_Customize_Manager $manager WP_Customize_Manager instance.
+	 * @param string $status The post status.
+	 * @return null|\WP_Error
+	 */
+	public function save( \WP_Customize_Manager $manager, $status = 'draft' ) {
+		$new_setting_ids = array_diff( array_keys( $this->post_data ), array_keys( $manager->settings() ) );
+		$manager->add_dynamic_settings( $new_setting_ids );
+
+		if ( ! empty( $this->contextual_post_data ) ) {
+			foreach ( $this->contextual_post_data as $setting_id => $contextual ) {
+				$this->snapshot->set_contextual( $setting_id, $contextual['value'], $contextual['dirty'] );
+			}
+		}
+
+		foreach ( $manager->settings() as $setting ) {
+			if ( $this->can_preview( $setting, $this->post_data ) ) {
+				$post_data = $this->post_data[ $setting->id ];
+				$this->snapshot->set( $setting, $post_data['value'], $post_data['dirty'] );
+			}
+		}
+
+		$r = $this->snapshot->save( $status );
+		if ( is_wp_error( $r ) ) {
+			status_header( 500 );
+			wp_send_json_error( $r->get_error_message() );
+		}
+	}
+
+	/**
+	 * Save snapshots via AJAX.
+	 *
+	 * Fires at `customize_save_after` to update and publish the snapshot.
+	 *
+	 * @param \WP_Customize_Manager $manager WP_Customize_Manager instance.
+	 */
+	public function save_snapshot( \WP_Customize_Manager $manager ) {
+		if ( ! current_user_can( 'customize' ) ) {
+			status_header( 403 );
+			wp_send_json_error( 'customize_not_allowed' );
+		} else if ( 'POST' !== $_SERVER['REQUEST_METHOD'] ) {
+			status_header( 405 );
+			wp_send_json_error( 'bad_method' );
+		}
+
+		$uuid = null;
+		if ( ! empty( $_POST['snapshot_uuid'] ) ) {
+			if ( empty( $this->post_data ) ) {
+				status_header( 400 );
+				wp_send_json_error( 'missing_snapshot_customized' );
+			} else {
+				$uuid = $_POST['snapshot_uuid'];
+			}
+		}
+
+		if ( $uuid && $this->snapshot->is_valid_uuid( $uuid ) ) {
+			$this->snapshot->set_uuid( $uuid );
+			$this->save( $manager, 'publish' );
+		}
+	}
+
+	/**
+	 * Update snapshots via AJAX.
+	 */
+	public function update_snapshot() {
+		if ( ! check_ajax_referer( self::AJAX_ACTION, 'nonce', false ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'bad_nonce' );
+		} else if ( ! current_user_can( 'customize' ) ) {
+			status_header( 403 );
+			wp_send_json_error( 'customize_not_allowed' );
+		} else if ( 'POST' !== $_SERVER['REQUEST_METHOD'] ) {
+			status_header( 405 );
+			wp_send_json_error( 'bad_method' );
+		} else if ( empty( $_POST['customize_snapshot_uuid'] ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'invalid_customize_snapshot_uuid' );
+		} else if ( empty( $_POST['scope'] ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'invalid_customize_snapshot_scope' );
+		} else if ( empty( $this->post_data ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'missing_snapshot_customized' );
+		} else if ( empty( $_POST['preview'] ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'missing_preview' );
+		}
```

Comment:
Most of that came from your transaction patch... So that's you flexing those `if/else` :muscle:'s.